ivadomed/models.py

```diff
@@ -1429,6 +1429,214 @@ def forward(self, x):
+        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
+                               padding=1, bias=True)
+        self.bn3 = nn.BatchNorm2d(planes)
+        self.conv3 = nn.Conv2d(planes, planes * 2, kernel_size=1, bias=True)
+        self.relu = nn.ReLU(inplace=True)
+        self.downsample = downsample
+        self.stride = stride
+
+    def forward(self, x):
+        residual = x
+
+        out = self.bn1(x)
+        out = self.relu(out)
+        out = self.conv1(out)
+
+        out = self.bn2(out)
+        out = self.relu(out)
+        out = self.conv2(out)
+
+        out = self.bn3(out)
+        out = self.relu(out)
+        out = self.conv3(out)
+
+        if self.downsample is not None:
+            residual = self.downsample(x)
+
+        out += residual
+
+        return out
+
+class Hourglass(nn.Module):
+    """Stacked Hourglass Networks.
+       Hourglass network is an auto-encoder decoder architecture for the task of pose estimation. In this architecture, 
+       Features are processed across all scales and consolidated to best capture the various spatial relationships associated with the pose.
+       code from: https://github.com/bearpaw/pytorch-pose/blob/master/pose/models/hourglass.py 
+    .. seealso::
+        Alejandro Newell et al. "Stacked Hourglass Networks for Human Pose Estimation."
+        Proceedings of the European Conference on Computer Vision. 2016.
+
+    Args:
+        block (nn.Module): Deep bottleneck structure 
+        num_blocks (int): number of blocks in the encoder structure
+        depth (int): hourglass depth
+        name (str): model's name used for call in configuration file.
+    """    
+    def __init__(self, block, num_blocks, planes, depth):
+        super(Hourglass, self).__init__()
+        self.depth = depth
+        self.block = block
+        self.hg = self._make_hour_glass(block, num_blocks, planes, depth)
+
+    def _make_residual(self, block, num_blocks, planes):
+        layers = []
+        for i in range(num_blocks):
+            layers.append(block(planes*block.expansion, planes))
+        return nn.Sequential(*layers)
+
+    def _make_hour_glass(self, block, num_blocks, planes, depth):
+        branch = 3
+        hg = []
+        for i in range(depth):
+            res = []
+            for j in range(branch):
+                res.append(self._make_residual(block, num_blocks, planes))
+            if i == 0:
+                res.append(self._make_residual(block, num_blocks, planes))
+            hg.append(nn.ModuleList(res))
+        return nn.ModuleList(hg)
+
+    def _hour_glass_forward(self, n, x):
+        up1 = self.hg[n-1][0](x)
+        low1 = F.max_pool2d(x, 2, stride=2)
+        low1 = self.hg[n-1][1](low1)
+
+        if n > 1:
+            low2 = self._hour_glass_forward(n-1, low1)
+        else:
+            low2 = self.hg[n-1][3](low1)
+        low3 = self.hg[n-1][2](low2)
+        up2 = F.interpolate(low3, scale_factor=2)
```

Comment:
After training the `HourglassNet`, I encountered an error during conversion from `.pt` to `.onnx`:

<details>
<summary> Full traceback </summary>

```python
Traceback (most recent call last):
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/onnx/utils.py", line 483, in _export
    graph, params_dict, torch_out = _model_to_graph(model, args, verbose,
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/onnx/utils.py", line 334, in _model_to_graph
    graph, torch_out = _trace_and_get_graph_from_model(model, args, training)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/onnx/utils.py", line 291, in _trace_and_get_graph_from_model
    torch.jit._get_trace_graph(model, args, _force_outplace=False, _return_inputs_states=True)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/jit/__init__.py", line 279, in _get_trace_graph
    outs = module(*args, **kwargs)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 550, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/jit/__init__.py", line 358, in forward
    graph, out = torch._C._create_graph_by_tracing(
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/jit/__init__.py", line 349, in wrapper
    outs.append(self.inner(*trace_inputs))
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 548, in __call__
    result = self._slow_forward(*input, **kwargs)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 534, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1641, in forward
    y = self.hg[i](x)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 548, in __call__
    result = self._slow_forward(*input, **kwargs)
  File "/home/joshua/repos/ivadomed/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 534, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1551, in forward
    return self._hour_glass_forward(self.depth, x)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1540, in _hour_glass_forward
    low2 = self._hour_glass_forward(n-1, low1)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1540, in _hour_glass_forward
    low2 = self._hour_glass_forward(n-1, low1)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1540, in _hour_glass_forward
    low2 = self._hour_glass_forward(n-1, low1)
  File "/home/joshua/repos/ivadomed/ivadomed/models.py", line 1547, in _hour_glass_forward
    out = up1 + up2
RuntimeError: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension 3
```

</details>

After investigating, I think this is a bug in the original model architecture, so I have reported the bug (plus a potential fix) upstream to https://github.com/bearpaw/pytorch-pose/issues/124.

I am going to try retraining the model using this fix, and see if it lets me convert to `.onnx` without failure.